src/CFlatToWasm.ml

```diff
@@ -9,62 +9,143 @@ module W = Wasm
+  let debug = fun i -> dummy_phrase W.Ast.({
+    module_name = name_of_string "Karamel";
+    item_name = name_of_string "debug";
+    idesc = dummy_phrase (FuncImport (mk_var i))
+  })
+
+  (** not exposed in WasmSupport.fst, no fstar-compatible type. *)
+  let debug_type = dummy_phrase (W.Types.FuncType ([], []))
+end
+
+(******************************************************************************)
+(* Global State                                                               *)
+(******************************************************************************)
+
+(* For each module that has been code-gen'd, we register all of its public
+   exports in this table so that subsequent modules can lookup functions by name
+   and register corresponding import requests, on an as-needed basis. *)
+let imports: (string, [
+    | `Function of (int -> W.Ast.import) * W.Types.func_type W.Source.phrase
+    | `Global of W.Ast.import
+  ]) Hashtbl.t
+=
+  let imports = Hashtbl.create 41 in
+  Hashtbl.add imports "data_start" (`Global Base.data_start);
+  Hashtbl.add imports "debug" (`Function (Base.debug, Base.debug_type));
+  imports
 
 (******************************************************************************)
 (* Environments                                                               *)
 (******************************************************************************)
 
 type env = {
-  funcs: int StringMap.t;
-    (** Mapping each function to its index in the function index space. *)
-  globals: int StringMap.t;
-    (** Mapping each global to its index in the global index space. *)
-  n_args: int;
-    (** The number of arguments to the current function. Needed to compute the
-     * index of the four "scratch" variables that each function has (see dup32). *)
+  funcs: (string, int) Hashtbl.t;
+    (** Mapping each function to its index in the function index space. This may
+        be extended on the fly to account for new imports from other modules. *)
+
+  globals: (string, int) Hashtbl.t;
+    (** Mapping each global to its index in the global index space. Also may be
+        extended on the fly for the same reasons. *)
+
   strings: (string, int) Hashtbl.t;
     (** Mapping constant string literals to their offset relative to the start
      * of THIS module's data segment. *)
   data_size: int ref;
     (** The current size of THIS module's data segment. This field and the one
      * above are mutable, so as to lazily allocate string literals as we hit
      * them. *)
+
+  n_args: int;
+    (** The number of arguments to the current function. Needed to compute the
+     * index of the four "scratch" variables that each function has (see dup32). *)
   location: loc list;
     (** For debugging *)
 }
 
-let empty = {
-  funcs = StringMap.empty;
-  globals = StringMap.empty;
-  n_args = 0;
+let empty () = {
+  funcs = Hashtbl.create 41;
+
+  globals = Hashtbl.create 41;
+
   strings = Hashtbl.create 41;
```

Comment:
note that this was sharing the hash table across all modules, meaning each data segment was accumulating the previous modules' data

Reply:
Why 41 for the `Hashtbl` initial size ?

Reply:
https://stackoverflow.com/a/16724713/873567

"A prime number is good because it tends to spread values throughout the table."

Don't know if it makes much difference in practice, but it's a habit I have had for a while...